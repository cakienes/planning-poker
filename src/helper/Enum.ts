export enum UserStoryStatusEnum {
    VOTED = 'VOTED',
    ACTIVE = 'ACTIVE',
    NOT_VOTED = 'NOT_VOTED',
}
