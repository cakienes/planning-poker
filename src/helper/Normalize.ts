const maxLength: any = (max: number) => (value: string) => (value && value.length > max ? value.slice(0, max) : value);

export { maxLength };
