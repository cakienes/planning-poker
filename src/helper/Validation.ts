const required: any = (value: any) => (value ? undefined : 'Please fill');

const onlyNumber: any = (value: any) => {
    return value && value !== '0' && parseInt(value, 10) ? undefined : 'Must be a number';
};

export { required, onlyNumber };
