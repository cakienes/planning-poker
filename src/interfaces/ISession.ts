import IUserStory from './IUserStory';

export default interface ISession {
    id: string;
    sessionName: string;
    numberOfVoters: number;
    userStories: IUserStory[];

    optionalUrl?: string;
    developers: string[];
}
