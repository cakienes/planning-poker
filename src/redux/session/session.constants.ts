export const GET_SESSIONS_FROM_LOCAL_STORAGE = 'GET_SESSIONS_FROM_LOCAL_STORAGE';
export const ADD_SESSION = 'ADD_SESSION';
export const SET_SELECTED_SESSION = 'SET_SELECTED_SESSION';
export const GIVE_STORY_POINT = 'GIVE_STORY_POINT';
export const SET_NEW_DEVELOPER = 'SET_NEW_DEVELOPER';
export const SET_DEVELOPER = 'SET_DEVELOPER';
