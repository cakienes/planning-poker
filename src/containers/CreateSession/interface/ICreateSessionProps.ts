import { RouteComponentProps } from 'react-router';
import ISession from '../../../interfaces/ISession';

export default interface ICreateSessionProps extends RouteComponentProps<any> {
    createSession: (values: ISession) => void;
}
