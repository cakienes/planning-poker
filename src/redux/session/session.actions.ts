import ISession from '../../interfaces/ISession';
import {
    ADD_SESSION,
    GET_SESSIONS_FROM_LOCAL_STORAGE,
    GIVE_STORY_POINT,
    SET_DEVELOPER,
    SET_FINAL_SCORE,
    SET_NEW_DEVELOPER,
    SET_SELECTED_SESSION,
    SET_SELECTED_SESSION_BY_URL,
} from './session.constants';

export const getSessionsFromLocalStorage = () => (dispatch: Function) => {
    dispatch({
        type: GET_SESSIONS_FROM_LOCAL_STORAGE,
    });
};

export const createSession = (session: ISession) => (dispatch: Function) => {
    dispatch({
        type: ADD_SESSION,
        session,
    });
};
export const setSelectedSession = (id: string) => (dispatch: Function) => {
    dispatch({
        type: SET_SELECTED_SESSION,
        id,
    });
};

export const giveStoryPoint = (voterName: string, storyPoint?: number) => (dispatch: Function) => {
    dispatch({
        type: GIVE_STORY_POINT,
        voterName,
        storyPoint,
    });
};

export const setNewDeveloper = (guid: string) => (dispatch: Function) => {
    dispatch({
        type: SET_NEW_DEVELOPER,
        guid,
    });
};
export const setDeveloper = (developerId: string) => (dispatch: Function) => {
    dispatch({
        type: SET_DEVELOPER,
        developerId,
    });
};
export const setFinalScore = (finalScore: string) => (dispatch: Function) => {
    dispatch({
        type: SET_FINAL_SCORE,
        finalScore,
    });
};
export const setSelecedSessionByUrl = (optionalUrl: string) => (dispatch: Function) => {
    dispatch({
        type: SET_SELECTED_SESSION_BY_URL,
        optionalUrl,
    });
};
