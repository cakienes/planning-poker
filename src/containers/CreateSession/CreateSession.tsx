import React from 'react';
import { connect } from 'react-redux';
import { toastr } from 'react-redux-toastr';
import { withRouter } from 'react-router';
import uuid from 'uuid';
import { UserStoryStatusEnum } from '../../helper/Enum';
import ISession from '../../interfaces/ISession';
import IUserStory from '../../interfaces/IUserStory';
import { createSession } from '../../redux/session/session.actions';
import Layout from '../Layout/Layout';
import CreateSessionForm from './CreateSessionForm/CreateSessionForm';
import ICreateSessionProps from './interface/ICreateSessionProps';
export class CreateSession extends React.Component<ICreateSessionProps, any> {
    render() {
        return (
            <Layout>
                <CreateSessionForm onSubmit={this.onSubmit} />
            </Layout>
        );
    }

    onSubmit = (values: any): void => {
        if (values) {
            const userStories: IUserStory[] = [];
            // eslint-disable-next-line
            values.userStories.split('\n').map((storyName: string, index: number) => {
                if (storyName && storyName !== '') {
                    userStories.push({
                        storyName,
                        status: index === 0 ? UserStoryStatusEnum.ACTIVE : UserStoryStatusEnum.NOT_VOTED,
                    });
                }
            });

            const newSession: ISession = {
                id: uuid(),
                numberOfVoters: values.numberOfVoters,
                sessionName: values.sessionName,
                userStories,
            };

            this.props.createSession(newSession);
            this.props.history.push(`/session/${newSession.id}/scrum-master`);
        } else {
            toastr.error('Validation Error', 'Please fill the form');
        }
    };
}

export const mapDispatchToProps = (dispatch: Function) => {
    return {
        createSession: (values: ISession) => {
            dispatch(createSession(values));
        },
    };
};

export default withRouter(
    connect(
        null,
        mapDispatchToProps,
    )(CreateSession),
);
