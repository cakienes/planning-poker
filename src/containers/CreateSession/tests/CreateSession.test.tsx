import React from 'react';
import { toastr } from 'react-redux-toastr';
import { createRenderer, ShallowRenderer } from 'react-test-renderer/shallow';
import uuid from 'uuid';
import { UserStoryStatusEnum } from '../../../helper/Enum';
import ISession from '../../../interfaces/ISession';
import { createSession } from '../../../redux/session/session.actions';
import { CreateSession, mapDispatchToProps } from '../CreateSession';
import ICreateSessionProps from '../interface/ICreateSessionProps';

jest.mock('uuid', () => {
    return jest.fn().mockImplementation(() => {
        return 1;
    });
});

jest.mock('react-redux-toastr', () => {
    return {
        toastr: {
            error: jest.fn(),
        },
    };
});

describe('CreateSession.tsx', () => {
    let renderer: ShallowRenderer;
    const createSessionMock = jest.fn();
    const dumData: any = '';

    let appProps: ICreateSessionProps = {
        createSession: createSessionMock,
        history: {
            push: jest.fn(),
        } as any,
        location: dumData,
        match: dumData,
        staticContext: dumData,
    };

    beforeEach(() => {
        renderer = createRenderer();
        appProps = {
            createSession: createSessionMock,
            history: dumData,
            location: dumData,
            match: dumData,
            staticContext: dumData,
        };
    });

    it('renders correctly CreateSession', async () => {
        expect(renderer.render(<CreateSession {...appProps} />)).toMatchSnapshot();
    });

    it('onSubmit of CreateSession while values is not null', async () => {
        appProps = {
            ...appProps,
            history: {
                ...appProps.history,
                push: jest.fn(),
            },
        };
        renderer.render(<CreateSession {...appProps} />);
        const instance: any = renderer.getMountedInstance();

        const dumSessionValues: any = {
            numberOfVoters: 2,
            optionalUrl: undefined,
            sessionName: 'sessionName',
            userStories: 'userStories1\nUserStories2',
        };

        const dumSession: ISession = {
            numberOfVoters: 2,
            optionalUrl: undefined,
            sessionName: 'sessionName',
            developers: [],
            id: uuid(),
            userStories: [
                { storyName: 'userStories1', status: UserStoryStatusEnum.ACTIVE },
                { storyName: 'UserStories2', status: UserStoryStatusEnum.NOT_VOTED },
            ],
        };

        instance.onSubmit(dumSessionValues);

        expect(createSessionMock).toHaveBeenCalledWith(dumSession);
        expect(appProps.history.push).toHaveBeenCalledWith(`/session/${dumSession.id}/scrum-master`);
    });

    it('onSubmit of CreateSession while values is null', async () => {
        renderer.render(<CreateSession {...appProps} />);
        const instance: any = renderer.getMountedInstance();

        instance.onSubmit(null);
        expect(toastr.error).toHaveBeenCalledWith('Validation Error', 'Please fill the form');
    });

    describe('mapDispatchToProps of CreateSession', () => {
        it('should dispatch getLinksFromLocalStorage when called', () => {
            const dispatchMock = jest.fn();
            const result = mapDispatchToProps(dispatchMock);

            const dumSession: ISession = {
                numberOfVoters: 2,
                optionalUrl: undefined,
                sessionName: 'sessionName',
                developers: [],
                id: uuid(),
                userStories: [
                    { storyName: 'userStories1', status: UserStoryStatusEnum.ACTIVE },
                    { storyName: 'UserStories2', status: UserStoryStatusEnum.NOT_VOTED },
                ],
            };

            result.createSession(dumSession);

            expect(JSON.stringify(dispatchMock.mock.calls[0][0])).toEqual(JSON.stringify(createSession(dumSession)));
        });
    });
});
