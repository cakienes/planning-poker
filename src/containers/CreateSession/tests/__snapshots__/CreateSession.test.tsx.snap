// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSession.tsx renders correctly CreateSession 1`] = `
<Layout>
  <ForwardRef
    destroyOnUnmount={true}
    enableReinitialize={false}
    forceUnregisterOnUnmount={false}
    form="createSessionForm"
    getFormState={[Function]}
    keepDirtyOnReinitialize={false}
    onSubmit={[Function]}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
  />
</Layout>
`;
